user/what-is-a-workspace.rst

```diff
@@ -0,0 +1,314 @@
+The file ``stdout_stderr.log`` shows stdout and stderr interleaved as they were emitted.
+
+..
+	TODO(sloretz) what is streams.log?
+
+
+The ``log`` directory contains logs from building and testing packages.
+We've only built ``foo``, so there are only build logs.
+Let's add tests to ``foo`` and see the output.
+
+Make a new file for the test.
+
+.. code-block:: bash
+
+	touch src/foo/test_foo.py
+
+Put the following content into ``test_foo.py``:
+
+.. code-block:: python
+
+	import foo
+
+	def test_foo():
+	    assert foo.foo_func()
+
+
+Tell ``colcon`` to run the tests.
+
+.. code-block:: bash
+
+	# Make sure you run this command from the root of the workspace!
+	colcon test
+
+Lets look in the ``log`` directory again.
+
+::
+
+	log
+	├── build_2022-05-20_11-50-03/...
+	├── COLCON_IGNORE
+	├── latest -> latest_test
+	├── latest_build -> build_2022-05-20_11-50-03
+	├── latest_test -> test_2022-05-20_11-50-05
+	└── test_2022-05-20_11-50-05
+	    ├── events.log
+	    ├── foo
+	    │	├── command.log
+	    │	├── stderr.log
+	    │	├── stdout.log
+	    │	├── stdout_stderr.log
+	    │	└── streams.log
+	    └── logger_all.log
+
+
+A new symlink ``latest_test`` was created, and it points to a new directory ``log/test_<date and time>``.
+This holds the console output from running the test.
+Checkout ``stdout_stderr.log``  and see the output of the latest test
+
+.. code-block:: bash
+
+	cat log/latest_test/foo/stdout_stderr.log
+
+.. note::
+
+	The test's output can be printed to the console in real time with
+	``colcon test --event-handlers console_direct+``.
+
+.. note::
+
+	Use the command ``colcon test-result`` to see a summary of test results on the console after tests have been run.
+
+
+
+Install artifacts
+*****************
+
+The last directory to talk about is the ``install`` directory.
+It contains both the installed software, and shell scripts that enable you to use it.
+This is sometimes called the **install space**.
+
+.. note::
+
+	You can change where packages are installed to with the ``--install-base`` option to ``colcon build``.
+
+Let's look inside.
+
+::
+
+	install
+	├── COLCON_IGNORE
+	├── foo/...
+	├── local_setup.[bash|bat|ps1|sh|zsh|...]
+	├── _local_setup_util_[sh|ps1|...].py
+	└── setup.[bash|bat|ps1|sh|zsh|...]
+
+The package ``foo`` was installed into the directory ``install/foo``.
+By default colcon builds an **isolated workspace**.
+That means every package is installed into its own directory.
+
+The shell scripts set environment variables that allow you to use the the software.
+You'll need to invoke the shell scripts, which is commonly called **sourcing a workspace**.
+
+.. note::
+
+	Always source a workspace from a different terminal than the one you used ``colcon build``.
+	Failure to do so can prevent colcon from detecting incorrect dependencies.
+
+..
+	TODO(sloretz) what's the difference between local_setup and setup?
```

Comment:
My understanding is that `setup.bash` invokes the shell scripts on this workspace and all underlays.  This is useful when you are sourcing this workspace from a "fresh" shell.  `local_setup.bash` only sources this workspace, assuming you have already sourced the underlays separately.  It's basically a minor performance optimization if you know you've already sourced the underlays.

Reply:
Thanks! described local_setup in 87e841e9dfd7bd756a0d69c9815c15524cc76f70